src/bluez/adapter/peripheral.rs

```diff
@@ -512,154 +472,126 @@ impl ApiPeripheral for Peripheral {
 
     fn discover_characteristics_in_range(
         &self,
-        start: u16,
-        end: u16,
+        _start: u16,
+        _end: u16,
     ) -> Result<Vec<Characteristic>> {
-        let mut results = vec![];
-        let mut start = start;
-        loop {
-            debug!("discovering chars in range [{}, {}]", start, end);
-
-            let mut buf = att::read_by_type_req(start, end, B16(GATT_CHARAC_UUID));
-            let data = self.request_raw(&mut buf)?;
-
-            match att::characteristics(&data) {
-                Ok(result) => {
-                    match result.1 {
-                        Ok(chars) => {
-                            debug!("Chars: {:#?}", chars);
-
-                            // TODO this copy can be removed
-                            results.extend(chars.clone());
-
-                            if let Some(ref last) = chars.iter().last() {
-                                if last.start_handle < end - 1 {
-                                    start = last.start_handle + 1;
-                                    continue;
-                                }
-                            }
-                            break;
-                        }
-                        Err(err) => {
-                            // this generally means we should stop iterating
-                            debug!("got error: {:?}", err);
-                            break;
-                        }
-                    }
-                }
-                Err(err) => {
-                    error!("failed to parse chars: {:?}", err);
-                    return Err(Error::Other(format!(
-                        "failed to parse characteristics response {:?}",
-                        err
-                    )));
-                }
-            }
-        }
-
-        // fix the end handles (we don't get them directly from device, so we have to infer)
-        for i in 0..results.len() {
-            (*results.get_mut(i).unwrap()).end_handle =
-                results.get(i + 1).map(|c| c.end_handle).unwrap_or(end);
+        let (ref lock, ref cvar) = *self.state;
+        trace!("Waiting for all services to be resolved");
+        let _guard = cvar
+            .wait_while(lock.lock().unwrap(), |b| *b == PeripheralState::Connected)
+            .unwrap();
+
+        if *_guard == PeripheralState::NotConnected {
+            return Err(Error::NotConnected);
         }
 
-        // update our cache
-        let mut lock = self.characteristics.lock().unwrap();
-        results.iter().for_each(|c| {
-            lock.insert(c.clone());
-        });
+        debug!("All services are now resolved!");
 
-        Ok(results)
+        Ok(self
+            .characteristics
+            .lock()
+            .unwrap()
+            .clone()
```

Comment:
Very nit: You *might* be able to use cloned() after your map (and use refs up to that point), save the copy then discard? But it honestly doesn't matter much in this context, and I can't remember off the top of my head where cloned() comes in.